### PR TITLE
My imagery data source

### PR DESCRIPTION
**Pull request by odinkomnogim**

### CODE REVIEW

File: mapflow/mapflow.py [hunk 1]
```diff
@@ -1961,7 +2007,6 @@ def create_processing(self) -> None:
                        icon=QMessageBox.Warning)
             return
 
-        provider_index = self.dlg.providerIndex()
```
Comment (odinkomnogim) [inline mapflow/mapflow.py:1964]: Это нигде в функции не использовалось дальше, решила убрать (?)

Comment (AlexeyTrekin) [reply mapflow/mapflow.py:1964]: ok!

File: mapflow/mapflow.py [hunk 2]
```diff
@@ -802,7 +809,8 @@ def toggle_processing_checkboxes(self, raster_source: Union[QgsRasterLayer, str,
         :param raster_source: Provider name or None, depending on the signal, if one of the
             tile providers, otherwise the selected raster layer
         """
-        enabled = isinstance(raster_source, QgsRasterLayer)
+        provider = self.providers[self.dlg.providerIndex()]
+        enabled = isinstance(raster_source, QgsRasterLayer) or isinstance(provider, MyImageryProvider)
```
Comment (AlexeyTrekin) [inline mapflow/mapflow.py:813]: See general comment

**Review by AlexeyTrekin — COMMENTED** [review]
I would prefer to change the providerCombo from qgsLayerCombo to its superclass qComboBox, because we don't need raster layers here anymore. It will make things easier, because no need to manage hide/show layers.
But we'll need to check for every providerCombo entry in the code to carefully see that it is handled correctly as qComboBox and all "layer-related" properties are removed

File: mapflow/entity/provider/default.py [hunk 3]
```diff
@@ -79,6 +79,34 @@ def is_default(self):
...
+    
+    @property
+    def requires_image_id(self):
+        return False
+    
+    @classmethod
+    def to_processing_params(self,
+                             image_id: Optional[str] = None,
+                             provider_name: Optional[str] = None):
+        return PostProviderSchema(data_provider=provider_name), {}
```
[the remark this answers is not in the document]
Comment (odinkomnogim) [reply mapflow/entity/provider/default.py:107]: Работает, но переделала на передачу url и PostSourceSchema

File: mapflow/mapflow.py [hunk 4]
```diff
@@ -1590,10 +1601,35 @@ def calculate_aoi_area_use_image_extent(self, use_image_extent: bool) -> None:
...
+                aoi = QgsGeometry().fromWkt(image.footprint)
+                self.calculate_aoi_area(aoi, helpers.WGS84)
+            elif mosaic and not image:
+                self.http.get(url=mosaic.rasterLayer.tileJsonUrl, callback=self.get_aoi_from_tileJson)
+            else:
+                self.dlg.disable_processing_start(reason=self.tr('Choose mosaic or image to start processing'),
+                                                  clear_area=True)
+                self.aoi = self.aoi_size = None
+    
+    def get_aoi_from_tileJson(self, response: QNetworkReply):
```
[the remark this answers is not in the document]
Comment (odinkomnogim) [reply mapflow/mapflow.py:1629]: I tried creating a temporary layer and adding footprints of all images in a mosaic to it as features... But that's obviosly not a solution. 
Something like 'mosaic.footprint' would be great to have! But untill we have it - I will leave it like this, I assume

File: mapflow/mapflow.py [hunk 5]
```diff
@@ -1916,6 +1920,20 @@ def get_aoi(self,
...
                 # raise PluginError(self.tr("Please select image in Search table for {}").format(provider.name))
+            elif isinstance(provider, MyImageryProvider):
+                image = self.data_catalog_service.selected_image()
+                mosaic = self.data_catalog_service.selected_mosaic()
+                if image:
+                    catalog_aoi = QgsGeometry().fromWkt(image.footprint)
+                elif mosaic:
+                    #catalog_aoi = self.http.get(url=mosaic.rasterLayer.tileJsonUrl, callback=self.get_aoi_from_tileJson)
+                    #wait untill mosaic.footprint (API), because previous line with callback does not work !!!
+                    catalog_aoi = selected_aoi
```
Comment (odinkomnogim) [inline mapflow/mapflow.py:1931]: temporary

File: mapflow/functional/api/data_catalog_api.py [hunk 6]
```diff
@@ -126,10 +126,13 @@ def add_mosaic_with_extent(self,
     def upload_image(self,
                      mosaic_id: UUID,
                      image_path: Union[Path, str],
+                     content_length: int,
```
Comment (odinkomnogim) [inline mapflow/functional/api/data_catalog_api.py:129]: Did I get "send Content-Length: <image size in bytes> header in every upload image request" right?

File: mapflow/functional/service/data_catalog.py [hunk 7]
```diff
@@ -136,37 +139,64 @@ def upload_images_to_mosaic(self):
...
+                                                 " MB").format(size=self.image_max_size_pixels,
+                                                               memory=self.image_max_size_bytes // (1024 * 1024)))
+            # Check if user has enougth stogage
+            content_length=Path(image_to_upload).stat().st_size
+            if content_length > self.free_storage:
+                message = (self.tr("<b>Not enough storage space. </b>"
+                                   "You have {free_storage} MB left, but '{name}' is "
+                                   "{content_length} MB".format(free_storage=round(self.free_storage/(1024*1024), 1),
+                                                                name=Path(image_to_upload).name,
+                                                                content_length=round(content_length/(1024*1024), 1)
```
Comment (AlexeyTrekin) [inline mapflow/functional/service/data_catalog.py:175]: Are you sure that we pass it as MB to the API?

File: mapflow/functional/service/data_catalog.py [hunk 8]
```diff
@@ -136,37 +139,64 @@ def upload_images_to_mosaic(self):
...
                 self.api.upload_image_error_handler(image_paths=failed)
             self.get_mosaic(mosaic_id)
             self.get_mosaic_images(mosaic_id)
+            self.mosaicsUpdated.emit()
         else:
             image_to_upload = image_paths[0]
-            non_uploaded = image_paths[1:] 
+            non_uploaded = image_paths[1:]
+            # Check if raster to be uploaded meets restrictions
+            layer = QgsRasterLayer(image_to_upload, "rasterLayerCheck", 'gdal')
```
Comment (AlexeyTrekin) [inline mapflow/functional/service/data_catalog.py:161]: We've discussed to use GDAL, not layer here to get the image stats

File: mapflow/functional/api/data_catalog_api.py (hunk 6 above)
Comment (AlexeyTrekin) [reply mapflow/functional/api/data_catalog_api.py:129]: Yes, you should do it in every request, if it is possible.

**Pull request merged**